feat: component Impersonate for en.admin.impersonate_component

KarlHeitmann opened the pull request:
Created component to send the impersonate request to the controller
through the users list of the admin: https://github.com/joemasilotti/railsdevs.com/issues/606

<!-- Before you submit a pull request for review, please make sure... -->

- [ ] My code contains tests covering the code I modified
- [X] I linted and tested the project with `bin/check`
- [ ] I added significant changes and product updates to the [changelog](CHANGELOG.md)

<!-- If this PR is not ready for review, please make sure to submit it as a draft. -->

joemasilotti commented:
Hey @KarlHeitmann, is this PR ready for review? If not, what are next steps? Let me know if you need any help getting it over the finish line!

KarlHeitmann commented:
Hi @joemasilotti , technically it is ready. But I don't know if it is necesarry to check the missing checkboxes:

[ ]  My code contains tests covering the code I modified
[ ] I added significant changes and product updates to the [changelog](https://github.com/joemasilotti/railsdevs.com/pull/CHANGELOG.md)

If it is necessary to complete these items, I'll need some help. How can I test what I did? What are the significant parts of the code you want to be tested? I think there may be some examples in the repo I can review, right?

About the second item,  I've never updated a Changelog before, so I don't know quite what to do.

I am keeping an eye on this thread.

Best,

Karl.

joemasilotti commented on app/components/admin/impersonate_component.rb:
```diff
@@ -0,0 +1,10 @@
+module Admin
+  class ImpersonateComponent < ApplicationComponent
+    attr_reader :email
+
+    def initialize(email)
+      super
+      @email = email
+    end
+  end
+end
```
I don't think we need a special component just for this. I also don't see how we would reuse this one as it's named a little specific.

I'd rather see something similar to `Admin::BasicLinkComponent` that takes in the form URL and an array of hidden fields. That way you could pass in the `:email` field. Also, the styling can match `Admin::BasicLinkComponent`.

joemasilotti reviewed (changes requested):
I made a comment/request on the new View Component you added.

Along with that, the tests on CI are failing because there are unused translations that need to be deleted. You can run `bundle exec i18n-tasks health en` locally to reproduce the error message.

Let me know if you need help creating the new component! I can also take care of the tests once that is ready.

KarlHeitmann commented:
Thanks for the feedback @joemasilotti , I can deal with the changelog and running the tests locally. I'll iterate over the component applying your feddback, and I'll show it to you. I think this weekend I will finish it.

Best

joemasilotti commented:
Sounds great, thanks @KarlHeitmann! Looking forward to it.

KarlHeitmann commented:
Hi @joemasilotti , tests are clean, changelog is OK. I've had the idea to rebase this branch on main, because you accepted PR #614 And did some other changes in september 8.

I've deleted the ImpersonateComponent, and created a new `Admin::BasicLinkPostComponent`, which creates a form with a hidden field, in the same style as `Admin::BasicLinkComponent`. The new `Admin::BasicLinkPostComponent` accepts only one pair: name/value, and this pair is written in the hidden input that is generated in the form.

If you want me to change something, squash some commits, or create a `fixup`, just tell me.

Best,

joemasilotti commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
 ## September
 
+* September 9 - feat: component Impersonate for en.admin.impersonate_component #627 @KarlHeitmann
```
```suggestion
* September 10 - Impersonate users directly from `/admin/users` #627 @KarlHeitmann
```

KarlHeitmann replied:
Done

joemasilotti commented on app/components/admin/basic_link_post_component.rb:
```diff
@@ -0,0 +1,18 @@
+  class BasicLinkPostComponent < ApplicationComponent
+    private attr_reader :title, :path, :name, :val
+
+    def initialize(title, path, name:, val:)
```
Let's rename `val` to `value` in the initializer and for the attribute. I don't see the need to remove a few characters.

KarlHeitmann replied:
yeah, I was about to use `value` but I forgot to.

joemasilotti commented on app/components/admin/basic_link_post_component.rb:
```diff
@@ -0,0 +1,18 @@
+    end
+
+    def call
+      form_with url: path do |form|
+        form.hidden_field(name, value: val) + form.submit(title, class: "hover:underline")
+      end
```
I think we can get by with `button_to`. Looking at the documentation again I'm noticing we can pass `params` as HTML options.

https://apidock.com/rails/v5.2.3/ActionView/Helpers/UrlHelper/button_to

I think something like this might work and be a bit cleaner.

```ruby
button_to(title, path, {class: "hover:underline", params: {name => val}})
```

KarlHeitmann replied:
yea, it works!!

joemasilotti reviewed (changes requested):
Looking good! I like this component much better as it enables us to reuse it somewhere for a similar scenario.

I made two code suggestions and one naming request. Once those are done we can merge this in!

KarlHeitmann commented:
I've got the changes on my local computer @joemasilotti  how do you want me to deliver the changes? send a fixup commit? create a new commit witth the fixes? or amend and then force - push?

joemasilotti commented:
Whatever works for you! I'll be squashing all of your commits into one merge commit, so no need to keep a tidy history.

KarlHeitmann commented:
Ready @joemasilotti , I'm not a big fan of modifying history, so I wrapped it up on a new commit. I can see it automatically changed the CHANGELOG.md file.

Sorry the delay, I missed the github notification of your message. I decided to take a look today and found your comment.

Best,

joemasilotti commented:
This looks great! Merging in now – thanks for your contribution, @KarlHeitmann.
